### ISSUE TITLE
panic client.PubObject with body length 0

### ISSUE THREAD

**Issue by gregf** [restic/restic#1058]
## Output of `restic version`

restic 0.6.1
compiled with go1.8.3 on linux/amd64

## How did you start restic exactly? (Include the complete command line)

restic backup /var/lib/vz/dump

## What backend/server/service did you use?

Amazon S3

## Expected behavior

To backup my data

## Actual behavior

I'm getting a panic while trying to backup 83GB of vm dumps to Amazon S3.

```
scan [/var/lib/vz/dump]                  
scanned 1 directories, 22 files in 0:00  
panic: client.PutObject: Put https://hugops-proxmox.s3.amazonaws.com/restic/data/03a04ac59d081bae84e4d78724936a253981364413d1268e179724549e726dd9: http: ContentLength=6960726 with Body length 0               

goroutine 50637 [running]:               
restic/archiver.(*Archiver).saveChunk(0xc420178690, 0x19761070f, 0x564733, 0x96dfcfea00000, 0xc437798000, 0x564733, 0x5da000, 0xc42009a790, 0xb76400, 0xc42051c030, ...)                                        
        src/restic/archiver/archiver.go:163 +0x270                                 
created by restic/archiver.(*Archiver).SaveFile                                    
        src/restic/archiver/archiver.go:237 +0x381                                 
using parent snapshot d2dcc97c 
```

## Steps to reproduce the behavior

restic backup /var/lib/vz/dump with a bunch of proxmox vms

## Do you have any idea what may have caused this?

Not sure. Maybe the size of the Objects?

```
$ cd /var/lib/vz/dumps
$ du -hs *

4.0K    vzdump-qemu-100-2017_06_03-08_39_42.log
5.4G    vzdump-qemu-100-2017_06_03-08_39_42.vma
4.0K    vzdump-qemu-100-2017_06_22-04_25_55.log
5.0G    vzdump-qemu-100-2017_06_22-04_25_55.vma.lzo
4.0K    vzdump-qemu-101-2017_06_03-08_42_49.log
8.4G    vzdump-qemu-101-2017_06_03-08_42_49.vma
4.0K    vzdump-qemu-101-2017_06_22-04_27_44.log
7.3G    vzdump-qemu-101-2017_06_22-04_27_44.vma.lzo
4.0K    vzdump-qemu-102-2017_06_03-08_48_34.log
3.3G    vzdump-qemu-102-2017_06_03-08_48_34.vma
4.0K    vzdump-qemu-102-2017_06_22-04_29_28.log
2.6G    vzdump-qemu-102-2017_06_22-04_29_28.vma.lzo
4.0K    vzdump-qemu-103-2017_06_03-08_50_33.log
26G     vzdump-qemu-103-2017_06_03-08_50_33.vma
8.0K    vzdump-qemu-103-2017_06_22-04_38_09.log
23G     vzdump-qemu-103-2017_06_22-04_38_09.vma.lzo
4.0K    vzdump-qemu-104-2017_06_03-09_00_03.log
4.0K    vzdump-qemu-104-2017_06_03-09_11_01.log
4.0K    vzdump-qemu-105-2017_06_22-04_48_19.log
2.4G    vzdump-qemu-105-2017_06_22-04_48_19.vma.lzo
4.0K    vzdump-qemu-107-2017_06_22-04_49_09.log
2.0G    vzdump-qemu-107-2017_06_22-04_49_09.vma.lzo
```

**Comment by fd0** [comment]
Thanks for leaving an issue. I've seen this error before, it is caused by S3 closing the connection before data could be sent. What's you upstream bandwidth to s3? Could you retry with with the master branch and `-o s3.connections=2`? Sometimes this error is caused by having too many open connections to s3 so that one connection starves and doesn't get any data through...

**Comment by gregf** [comment]
```
root@proxmox:/usr/local/bin# restic -o s3.connections=2 backup /var/lib/vz/dump
scan [/var/lib/vz/dump]
scanned 1 directories, 56 files in 0:00
panic: client.PutObject: invalid argument73.098 GiB  10 / 57 items  0 errors  ETA 38:55 
goroutine 1511 [running]:
restic/archiver.(*Archiver).saveChunk(0xc420286eb0, 0xc1e4e0, 0xc4200123d8, 0x6c4dba9, 0x477d00, 0x7aabe25f00000, 0xc42e8b2000, 0x477d00, 0x4ba000, 0xc420352b00, ...)
        src/restic/archiver/archiver.go:165 +0x395
created by restic/archiver.(*Archiver).SaveFile
        src/restic/archiver/archiver.go:239 +0x3a4

```

I have 1Gb up.

**Comment by gregf** [comment]
Another note this backup seems to always fail around the 70gb mark every time. 

Today I decided I would try B2 Cloud since that is what I wanted to use anyways and now I got master installed I could use it. 

I uploaded 213gb without any errors using B2 Cloud. I can still help debug with s3 though that data hasn't gone away yet.

**Comment by fd0** [comment]
Hm, very strange. I cannot reproduce it and I'm out of ideas what's the problem here. minio-go (the library we're using for accessing s3) has merged a [fix](https://github.com/minio/minio-go/issues/712) for a similar issue. I've created PR #1066, could you please test branch `update-minio-go` and retry?

**Comment by kamsz** [comment]
I encountered this issue today. It happens when I'm trying to do a first backup of a large directory (20+ GB).

**Comment by gregf** [comment]
@fd0 do I just pull & build restic again to get the new minio-go?

**Comment by fd0** [comment]
pull and checkout the branch with `git checkout update-minio-go`, then build restic with `go run build.go` and use `./restic` to run.

**Comment by gregf** [comment]
This seems to have fixed it for me at least. I ran it with the old version before I tried the new one. Four times in a row it died at the 70Gb mark. Tried the new binary uploaded 310Gb without fail. Thanks for the help!

**Comment by fd0** [comment]
I'm glad it helped, thanks for the feedback!

**Comment by fd0** [comment]
I've merged #1066, please leave a comment if this happens again to anybody. For now, I'm going to close this issue.
